gax/src/main/java/com/google/api/gax/batching/BatcherStats.java

```diff
@@ -0,0 +1,181 @@
+  private final Map<Code, Integer> entryStatusCounts = new HashMap<>();
+
+  /**
+   * Records the count of the exception and it's type when a complete batch failed to apply.
+   *
+   * <p>Note: This method aggregates all the subclasses of {@link ApiException} under ApiException
+   * using the {@link Code status codes} and its number of occurrences.
+   */
+  synchronized void recordBatchFailure(Throwable throwable) {
+    Class exceptionClass = throwable.getClass();
+
+    if (throwable instanceof ApiException) {
+      Code code = ((ApiException) throwable).getStatusCode().getCode();
+      exceptionClass = ApiException.class;
+
+      int oldStatusCount = MoreObjects.firstNonNull(requestStatusCounts.get(code), 0);
+      requestStatusCounts.put(code, oldStatusCount + 1);
+    }
+
+    int oldExceptionCount = MoreObjects.firstNonNull(requestExceptionCounts.get(exceptionClass), 0);
+    requestExceptionCounts.put(exceptionClass, oldExceptionCount + 1);
+  }
+
+  /**
+   * Records partial failures within each batch. partialBatchFailures counts the number of batches
+   * that have at least one failed entry while entryStatusCounts and entryExceptionCounts track the
+   * count of the entries themselves.
+   *
+   * <p>Note: This method aggregates all the subclasses of {@link ApiException} under ApiException
+   * using the {@link Code status codes} and its number of occurrences.
+   */
+  synchronized <T extends ApiFuture> void recordBatchElementsCompletion(
+      List<T> batchElementResultFutures) {
+    boolean isRequestPartiallyFailed = false;
+    for (final ApiFuture elementResult : batchElementResultFutures) {
+      try {
+        elementResult.get();
+      } catch (Throwable throwable) {
+
+        if (!isRequestPartiallyFailed) {
+          isRequestPartiallyFailed = true;
+          requestPartialFailureCount++;
+        }
+        Throwable actualCause = throwable.getCause();
+        Class exceptionClass = actualCause.getClass();
+
+        if (actualCause instanceof ApiException) {
+          Code code = ((ApiException) actualCause).getStatusCode().getCode();
+          exceptionClass = ApiException.class;
+
+          int oldExceptionCount = MoreObjects.firstNonNull(entryStatusCounts.get(code), 0);
+          entryStatusCounts.put(code, oldExceptionCount + 1);
+        }
+
+        int oldExceptionCount =
+            MoreObjects.firstNonNull(entryExceptionCounts.get(exceptionClass), 0);
+        entryExceptionCounts.put(exceptionClass, oldExceptionCount + 1);
+      }
+    }
+  }
+
+  /** Calculates and formats the message with request and entry failure count. */
+  @Nullable
+  synchronized BatchingException asException() {
+    if (requestExceptionCounts.isEmpty() && requestPartialFailureCount == 0) {
+      return null;
+    }
+
+    StringBuilder sb = new StringBuilder();
+    sb.append("Batching finished with ");
+
+    if (!requestExceptionCounts.isEmpty()) {
```

Comment:
This would be `false` when if only some of the entries are failed within a batching session Or to put it in another word, the batch is only partially failed.